File: src/http.ffi.mjs
import { readFileSync } from "node:fs";
import * as Fs from "node:fs/promises";
import * as Http from "node:http";
import * as Path from "node:path";
import * as Process from "node:process";

const cwd = Process.cwd();
const root = Path.join(cwd, "build/dev/javascript");
const toml = readFileSync(Path.join(cwd, "gleam.toml"), "utf-8");
const name = toml.match(/name *= *"(.+)"/)[1];

const html = `<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Lustre preview server</title>

  <script type="module">
    import { main } from "./${name}/${name}.mjs"

    document.addEventListener("DOMContentLoaded", () => {
      main();
    });
  </script>
</head>
<body>
  <div data-lustre-app></div>
</body>
</html>`;

const server = Http.createServer((req, res) => {
  switch (true) {
    case req.url === "/": {
      res.setHeader("Content-Type", "text/html");
      res.statusCode = 200;
      res.end(html);

      break;
    }

    case req.url.endsWith(".js"):
    case req.url.endsWith(".mjs"): {
      Fs.readFile(Path.join(root, req.url), "utf-8")
        .then((src) => {
          res.setHeader("Content-Type", "application/javascript");
          res.statusCode = 200;
          res.end(src);
        })
        .catch((_err) => {
          res.statusCode = 404;
          res.end();
        });

      break;
    }

    case req.url.endsWith(".css"): {
      Fs.readFile(Path.join(root, req.url), "utf-8")
        .then((src) => {
          res.setHeader("Content-Type", "text/css");
          res.statusCode = 200;
          res.end(src);
        })
        .catch((_err) => {
          res.statusCode = 404;
          res.end();
        });

      break;
    }

    default: {
      Fs.readFile(Path.join(root, req.url), "utf-8")
        .then((src) => {
          res.setHeader("Content-Type", "text/plain");
          res.statusCode = 200;
          res.end(src);
        })
        .catch((_err) => {
          res.statusCode = 404;
          res.end();
        });
    }
  }
});

export const serve = (host, port, on_start, on_port_taken) => {
  let tries = 1;
  server.on("error", (error) => {
    if (error.code === "EADDRINUSE") {
      let is_first_try = tries === 1;
      if (is_first_try) {
        on_port_taken(port);
      }
      tries++;
      port++;
      server.listen(port, host);
    }
  }).listen(port, host, () => { on_start(port) });
};

File: src/http_ffi.erl
-module(http_ffi).
-export([serve/3, response_default_headers/0]).

serve({options, Host, Port, IncludeStyles}, OnStart, OnPortTaken) ->
    {ok, Pattern} = re:compile("name *= *\"(?<Name>.+)\""),
    {ok, Toml} = file:read_file("gleam.toml"),
    {match, [Name]} = re:run(Toml, Pattern, [{capture, all_names, binary}]),

    Html =
        <<
            "<!DOCTYPE html>\n"
            "<html lang=\"en\">\n"
            "<head>\n"
            "  <meta charset=\"UTF-8\">\n"
            "  <meta name=\"viewport\" content=\"width=device-width, initial-scale=1.0\">\n"
            "  <title>Lustre preview server</title>\n",
            case IncludeStyles of
                true ->
                     <<"  <link rel=\"stylesheet\" href=\"https://cdn.jsdelivr.net/gh/lustre-labs/ui/priv/styles.css\">\n">>;
                false ->
                    <<"">>
            end/binary,
            "  <script type=\"module\">\n"
            "    import { main } from './",
            Name/binary,
            "/",
            Name/binary,
            ".mjs'\n"
            "\n"
            "    document.addEventListener(\"DOMContentLoaded\", () => {\n"
            "      main();\n"
            "    });\n"
            "  </script>\n"
            "</head>\n"
            "<body>\n"
            "  <div data-lustre-app></div>\n"
            "</body>\n"
            "</html>"
        >>,

    file:write_file("build/dev/javascript/index.html", Html),

    AbsPath =
        string:trim(
            filename:absname("build/dev/javascript"), trailing, "/."
        ),

    inets:start(),
    Address = {127, 0, 0, 1},

    ActualPort =
        case port_available(Port) of
            true ->
                Port;
            false ->
                OnPortTaken(Port),
                first_available_port(Port + 1)
        end,

    {ok, Pid} =
        httpd:start_service([
            {bind_address, Address},
            {document_root, AbsPath},
            {server_root, AbsPath},
            {directory_index, ["index.html"]},
            {server_name, binary_to_list(Host)},
            {port, ActualPort},
            {default_type, "text/html"},
            {mime_types, mime_types()},
            {customize, ?MODULE},
            {modules, [mod_alias, mod_dir, mod_get]}
        ]),

    OnStart(ActualPort),

    receive
        {From, shutdown} ->
            ok = httpd:stop_service(Pid),
            From ! done
    end.

port_available(Port) ->
    case gen_tcp:listen(Port, []) of
        {ok, Sock} ->
            ok = gen_tcp:close(Sock),
            true;
        _ ->
            false
    end.

first_available_port(Port) ->
    case port_available(Port) of
        true -> Port;
        false -> first_available_port(Port + 1)
    end.

mime_types() ->
    [
        {"html", "text/html"},
        {"htm", "text/html"},
        {"js", "text/javascript"},
        {"mjs", "text/javascript"},
        {"css", "text/css"},
        {"gif", "image/gif"},
        {"jpg", "image/jpeg"},
        {"jpeg", "image/jpeg"},
        {"png", "image/png"}
    ].

response_default_headers() ->
    [
        {"cache-control", "no-store, no-cache, must-revalidate, private"},
        {"pragma", "no-cache"}
    ].
